src/GraphQL/Utility/FileUpload.php

```diff
@@ -189,68 +195,233 @@ public function createTemporaryFileUpload(UploadedFile $file, array $settings):
+    $destination = $this->getUploadLocation($settings);
+
+    // Check the destination file path is writable.
+    if (!$this->fileSystem->prepareDirectory($destination, FileSystemInterface::CREATE_DIRECTORY)) {
+      $response->addViolation($this->t('Unknown error while uploading the file "@file".', ['@file' => $uploaded_file->getClientOriginalName()]));
+      $this->logger->error('Could not create directory "@upload_directory".', ["@upload_directory" => $destination]);
       return $response;
     }
-    $name = $file->getClientOriginalName();
-    $mime = $this->mimeTypeGuesser->guess($name);
-    $destination = $this->fileSystem->getDestinationFilename("{$uploadDir}/{$name}", $this->fileSystem::EXISTS_RENAME);
 
-    // Begin building file entity.
-    $values = [
-      'uid' => $this->currentUser->id(),
-      'status' => 0,
-      'filename' => $name,
-      'uri' => $destination,
-      'filesize' => $file->getSize(),
-      'filemime' => $mime,
-    ];
-    $storage = $this->entityTypeManager->getStorage('file');
-    /** @var \Drupal\file\FileInterface $fileEntity */
-    $fileEntity = $storage->create($values);
+    $validators = $this->getUploadValidators($settings);
 
-    // Validate the entity values.
-    $violations = $fileEntity->validate();
-    if ($violations->count()) {
-      foreach ($violations as $violation) {
-        $response->addViolation($violation->getMessage());
-      }
+    $prepared_filename = $this->prepareFilename($uploaded_file->getClientOriginalName(), $validators);
+
+    // Create the file.
+    $file_uri = "{$destination}/{$prepared_filename}";
+
+    $temp_file_path = $uploaded_file->getRealPath();
+
+    $file_uri = $this->fileSystem->getDestinationFilename($file_uri, FileSystemInterface::EXISTS_RENAME);
+
+    // Lock based on the prepared file URI.
+    $lock_id = $this->generateLockIdFromFileUri($file_uri);
+
+    if (!$this->lock->acquire($lock_id)) {
+      $response->addViolation($this->t('Unknown error while uploading the file "@file".', ['@file' => $uploaded_file->getClientOriginalName()]));
       return $response;
     }
 
-    // Validate the file name length.
-    if ($violations = file_validate($fileEntity, $this->getUploadValidators($settings))) {
-      $response->addViolations($violations);
+    // Begin building file entity.
+    /** @var \Drupal\file\FileInterface $file */
+    $file = $this->fileStorage->create([]);
+    $file->setOwnerId($this->currentUser->id());
+    $file->setFilename($prepared_filename);
+    $file->setMimeType($this->mimeTypeGuesser->guess($prepared_filename));
+    $file->setFileUri($file_uri);
+    // Set the size. This is done in File::preSave() but we validate the file
+    // before it is saved.
+    $file->setSize(@filesize($temp_file_path));
+
+    // Validate the file entity against entity-level validation and field-level
+    // validators.
+    if (!$this->validate($file, $validators, $response)) {
       return $response;
```

Comment:
shouldn't we release the lock before returning response?

Reply:
nice, you just found a bug in core's FileUploadResource which we copied :)